ssh-agency.el

```diff
@@ -136,8 +137,72 @@ ssh-agency always finds the agent without consulting this file."
   :type '(choice (repeat (file :must-match t))
                  (const nil :tag "ssh-add's default")))
 
+(defconst ssh-agency-socket-regexp
+  ;; Try to match
+  ;; - /tmp/ssh-XXXXX/agent.NNN (ssh-agent default)
+  ;; - /run/user/[uid]/keyring/ssh (gnome-keyring-daemon 1)
+  ;; - ~/.cache/keyring-XXXXX/ssh (gnome-keyring-daemon 2)
+  ;; But don't match
+  ;; - /run/user/[uid]/gnupg/S.gpg-agent.ssh
+  "/\\(?:agent[.][0-9]+\\|ssh\\)\\'")
+
+(defcustom ssh-agency-socket-locaters
+  `(,@(when (executable-find "ss")
+        `((ssh-agency-find-socket-from-ss
+           :glob "*ssh*" :regexp ,ssh-agency-socket-regexp)))
+    ,@(when     ; Windows has a netstat command, but it won't help us.
+          (and (not (eq system-type 'windows-nt))
+               (executable-find "netstat"))
+        `((ssh-agency-find-socket-from-netstat
+           :regexp ,ssh-agency-socket-regexp)))
+    ;; Look in ssh-agent's default location.
+    (ssh-agency-find-socket-from-glob
+     ,(concat temporary-file-directory "ssh-*/agent.*")))
+  "List of (FUN . ARGS) to search for ssh-agent sockets."
+  :group 'ssh-agency
+  :type '(alist :key-type function))
+
 ;;; Functions
 
+(defun ssh-agency-socket-active-p (socket)
+  "Return non-nil if SOCKET can be used to communicate with an ssh-agent."
+  (when (eq system-type 'windows-nt)
+    ;; Follow the lead of msysgit's start-ssh-agent.cmd:
+    ;; replace %TEMP% with "/tmp".
+    (setq socket (replace-regexp-in-string
+                  (concat "\\`" (regexp-quote temporary-file-directory))
+                  "/tmp/" socket)))
+  (let ((process-environment (cons (concat "SSH_AUTH_SOCK=" socket)
+                                   process-environment)))
+    (ssh-agency-status)))
+
+(cl-defun ssh-agency-find-socket-from-ss (&key glob regexp)
```

Comment:
Just a one question, why you use `cl-defun` instead of `defun`?

Reply:
for the `&key` feature.

Reply:
The emacs-lisp don't define it on `defun`? Sorry for the dumb question, I have more experience with common lisp than emacs lisp.

Reply:
Yeah, as far as I know, it's because Stallman doesn't like `&key` args, so it wasn't added to elisp originally. So now it's only in the cl-lib extension.

Reply:
Funny fact. Thanks for sharing it.